pkg/applyconfigurations/gen.go

```diff
@@ -0,0 +1,230 @@
+		return err
+	}
+
+	into.AddHelp(isCRDMarker,
+		markers.SimpleHelp("apply", "enables apply configuration generation for this type"))
+	into.AddHelp(
+		enableTypeMarker, markers.SimpleHelp("apply", "overrides enabling or disabling applyconfigurations generation for the type"))
+	into.AddHelp(
+		enablePkgMarker, markers.SimpleHelp("apply", "overrides enabling or disabling applyconfigurations generation for the package"))
+	into.AddHelp(
+		outputPkgMarker, markers.SimpleHelp("apply", "overrides the default output package for the applyconfigurations generation"))
+	return nil
+}
+
+func enabledOnPackage(col *markers.Collector, pkg *loader.Package) (bool, error) {
+	pkgMarkers, err := markers.PackageMarkers(col, pkg)
+	if err != nil {
+		return false, err
+	}
+	pkgMarker := pkgMarkers.Get(enablePkgMarker.Name)
+	if pkgMarker != nil {
+		return pkgMarker.(bool), nil
+	}
+	return false, nil
+}
+
+// enableOnType marks whether applyconfiguration generation is enabled for the type.
+func enabledOnType(info *markers.TypeInfo) bool {
+	if typeMarker := info.Markers.Get(enableTypeMarker.Name); typeMarker != nil {
+		return typeMarker.(bool)
+	}
+	return isCRD(info)
+}
+
+func outputPkg(col *markers.Collector, pkg *loader.Package) (string, error) {
+	pkgMarkers, err := markers.PackageMarkers(col, pkg)
+	if err != nil {
+		return "", err
+	}
+	pkgMarker := pkgMarkers.Get(outputPkgMarker.Name)
+	if pkgMarker != nil {
+		return pkgMarker.(string), nil
+	}
+	return "", nil
+}
+
+// isCRD marks whether the type is a CRD based on the +kubebuilder:resource marker.
+func isCRD(info *markers.TypeInfo) bool {
+	objectEnabled := info.Markers.Get(isCRDMarker.Name)
+	return objectEnabled != nil
+}
+
+func (d Generator) Generate(ctx *genall.GenerationContext) error {
+	headerFilePath := d.HeaderFile
+
+	if headerFilePath == "" {
+		tmpFile, err := os.CreateTemp("", "applyconfig-header-*.txt")
+		if err != nil {
+			return fmt.Errorf("failed to create temporary file: %w", err)
+		}
+		tmpFile.Close()
+
+		defer os.Remove(tmpFile.Name())
+
+		headerFilePath = tmpFile.Name()
+	}
+
+	objGenCtx := ObjectGenCtx{
+		Collector:      ctx.Collector,
+		Checker:        ctx.Checker,
+		HeaderFilePath: headerFilePath,
+	}
+
+	for _, pkg := range ctx.Roots {
+		if err := objGenCtx.generateForPackage(pkg); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ObjectGenCtx contains the common info for generating apply configuration implementations.
+// It mostly exists so that generating for a package can be easily tested without
+// requiring a full set of output rules, etc.
+type ObjectGenCtx struct {
+	Collector      *markers.Collector
+	Checker        *loader.TypeChecker
+	HeaderFilePath string
+}
+
+// generateForPackage generates apply configuration implementations for
+// types in the given package, writing the formatted result to given writer.
+// May return nil if source could not be generated.
```

Comment:
What does that mean? :) That we might not get an error if "source could not be generated"? (what is source in this context?)